enhancements/network/sdn-live-migration.md

```diff
@@ -0,0 +1,476 @@
+
+On the SDN side:
+- We have to change ingress NetworkPolicy processing to be based entirely on pod
+  IPs rather than using namespace VNIDs since packets from OVN nodes won't have
+  the VNID set. There is already code to generate the rules that way, though,
+  because egress NetworkPolicy already works that way.
+
+### Workflow Description
+
+#### Migration Setup
+
+1. The admin kicks off the migration process by updating the custom resource
+   `network.operator`:
+   
+   ```bash
+   $ oc patch Network.operator.openshift.io cluster --type='merge' --patch '{"spec":{"migration":{"networkType":"OVNKubernetes","mode":"Live"}}}"
+   ```
+   
+   CNO will check if there are any unsupported features (refer to the
+   [Limitations](#limitations) section) enabled in the cluster. If yes, it will
+   set an warning message under the `status.migration` field of the custom
+   resource `network.operator`.
+
+2. CNO will check the current hardware MTU and the cluster MTU to see if there
+   is enough free space for OVN-K overlay overhead. If not, CNO will be set to
+   degraded. Users must decrease the cluster MTU to the maximum value that can
+   be supported by OVN Kubernetes before conducting the migration. If yes, the
+   condition `Progressing` of the network Cluster Operator is switched from
+   `False` to `True`
+
+3. CNO will redeploy openshift-sdn in migration mode. It will add a condition
+   check to the wrapper script of the sdn container to see if the bridge `br-ex`
+   exists on the node. If `br-ex` exists, it means the node has already been
+   updated by MCO, therefore it's ready for running OVN Kubernetes. The sdn pod
+   would just sleep infinity rather than actually launching the
+   `openshift-sdn-node` process.
+
+4. CNO will also deploy OVN Kubernetes to the cluster with hybrid overlay
+   enabled. To avoid racing, the host subnet allocation will be disabled in
+   OVN Kubernetes.
+   
+   The ovnkube-node container wrapper script will be rendered with the logic for
+   migration:
+
+   - If `br-ex` doesn't exist, it means the node has not yet been updated by
+     MCO, so instead of starting the `ovnkube-node` process, it would run the
+     `hybrid-overlay-node` process. `hybrid-overlay-node` can add the node
+     annotation `k8s.ovn.org/hybrid-overlay-distributed-router-gateway-mac`,
+     which is required by hybrid overlay.
+
+     The script will also add the node annotation
+     `k8s.ovn.org/hybrid-overlay-node-subnet`, according to the HostSubnet CR of
+     this node.
+
+   - If `br-ex` does exist, it means the node has been updated by MCO. So the
+     `ovnkube-node` can be started.
+
+     The script will also add the node annotation `k8s.ovn.org/node-subnets`
+     according to the HostSubnet CR of this node.
+
+#### Migration
+
+When the condition 'Progressing' of the network Cluster Operator is switched
+from `True` to `False`, the admin can kick off the migration from CNO by:
```

Comment:
Similar to the MTU procedure, is there a reason a human needs to do a multi-step process here? Is there a reason we can't declaratively say "move to OVN" and all the steps are done by operators?

Reply:
The answer is similar to the above.